### PR TITLE
fix: 修复自定义目录树同名节点展示异常 & 导出缺失角头 close #2455

### LINKED ISSUE #2455: 🐛 自定义目录树的tooltip显示和折叠的rowKey有问题
### 🏷 Version

| Package        | Version |
| -------------- | ------- |
| @antv/s2       |  1.52.1  |
| @antv/s2-react |  1.44.3  |
| @antv/s2-vue   |         |

### Sheet Type

 ✅  PivotSheet
- [ ] TableSheet
- [ ] GridAnalysisSheet
- [ ] StrategySheet
- [ ] EditableSheet

### 🖋 Description

自定义目录树存在两个问题：
1. 如果两个树的父节点的`title`一样，那么一个折叠另一个也会折叠，而不是根据`key`来的
2. 默认tooltip展示的又是节点的`key`，而不是`title`

具体看截图，下面有codesandbox 链接
image
image

### 🔗 Reproduce Link

https://codesandbox.io/p/sandbox/old-currying-p3xz89?file=%2Findex.tsx
### 😊 Expected Behavior
希望折叠的根据`key`来做唯一标识，tooltip展示`title`

### 💻 System information

| Environment | Info |
| ------- | ------- |
| System  |         |
| Browser |         |

### PR DESCRIPTION
**Pull request by lijinke666**
### 👀 PR includes

🐛 Bugfix

- [x] Solve the issue and close #2455

### 📝 Description

- 自定义行列头使用 field 生成 id, 解决同名节点引起的交互/折叠/复制等衍生问题.

```diff
- "root[&]自定义节点A"
+ "root[&]custom-node-1"
```

- 兼容刷选和复制导出, 并完善相关单测.
- 修复自定义行头 (平铺/树状) 复制的数据未包含角头的问题.

### 🖼️ Screenshot

| Before | After |
| ------ | ----- |
| Kapture 2024-02-19 at 19 05 29 | Kapture 2024-02-19 at 19 03 34 |
| image | image |
| image | image |

### 🔍 Self-Check before the merge

- [ ] Add or update relevant docs.
- [ ] Add or update relevant demos.
- [ ] Add or update test case.
- [ ] Add or update relevant TypeScript definitions.

### CODE REVIEW

File: packages/s2-core/__tests__/spreadsheet/hidden-columns-spec.ts [hunk 1]
```diff
@@ -457,16 +457,13 @@ describe('SpreadSheet Hidden Columns Tests', () => {
...
-      const parentNode = pivotSheet.facet.getColNodeById('root[&]笔[&]义乌')!;
-
       const hiddenColumnsInfo = pivotSheet.store.get('hiddenColumnsDetail')[0];
 
       expect(rootNode.width).toEqual(100);
       expect(rootNode.x).toEqual(100);
       expect(grandTotalsNode.width).toEqual(100);
       expect(grandTotalsNode.x).toEqual(0);
       expect(hiddenColumnsInfo).toBeTruthy();
-      expect(parentNode.hiddenChildNodeInfo).toEqual(hiddenColumnsInfo);
```
Comment (lijinke666) [inline packages/s2-core/__tests__/spreadsheet/hidden-columns-spec.ts:469]: 1.x 已经移除了, 不再消费

File: packages/s2-core/src/utils/export/copy/pivot-data-cell-copy.ts [hunk 2]
```diff
@@ -235,8 +244,43 @@ export class PivotDataCellCopy extends BaseDataCellCopy {
...
+        const result: string[] = new Array(maxRowLen).fill('');
+
+        result[maxRowLen - 1] = node.value;
+
+        return result;
+      });
+    }
+
+    // 平铺模式
+    return Object.values(groupBy(sortedCornerNodes, 'y')).map((nodes) => {
```
Comment (lijinke666) [inline packages/s2-core/src/utils/export/copy/pivot-data-cell-copy.ts:268]: image

File: packages/s2-core/src/utils/export/copy/pivot-data-cell-copy.ts [hunk 3]
```diff
@@ -235,8 +244,43 @@ export class PivotDataCellCopy extends BaseDataCellCopy {
...
+  protected getCustomRowCornerMatrix = (rowMatrix?: string[][]): string[][] => {
+    const maxRowLen = getMaxRowLen(rowMatrix ?? []);
+    const cornerNodes = this.spreadsheet.facet.getCornerNodes();
+    // 对 cornerNodes 进行排序， cornerType === CornerNodeType.Col 的放在前面
+    const sortedCornerNodes = sortBy(cornerNodes, (node) => {
+      return node.cornerType === CornerNodeType.Col ? 0 : 1;
+    });
+
+    // 树状模式
+    if (this.spreadsheet.isHierarchyTreeType()) {
```
Comment (lijinke666) [inline packages/s2-core/src/utils/export/copy/pivot-data-cell-copy.ts:256]: image

File: packages/s2-react/src/components/export/index.tsx [hunk 4]
```diff
@@ -56,6 +56,8 @@ export const Export: React.FC<ExportProps> = React.memo((props) => {
 
   const PRE_CLASS = `${S2_PREFIX_CLS}-export`;
 
+  const [messageApi, contextHolder] = message.useMessage();
```
Comment (lijinke666) [inline packages/s2-react/src/components/export/index.tsx:59]: image

**Review by wjgogogo — APPROVED** [review]

**Pull request merged**